File: Sources/RepoBarCore/Models/RepositoryIdentity.swift
import Foundation

public struct RepositoryIdentity: Equatable, Codable, Hashable, Sendable {
    public var provider: SourceControlProvider
    public var id: String
    public var name: String
    public var namespacePath: String
    public var pathWithNamespace: String
    public var slug: String
    public var webURL: URL?
    public var apiURL: URL?
    public var providerSpecificID: String?

    public init(
        provider: SourceControlProvider,
        id: String,
        name: String,
        namespacePath: String,
        pathWithNamespace: String? = nil,
        slug: String? = nil,
        webURL: URL? = nil,
        apiURL: URL? = nil,
        providerSpecificID: String? = nil
    ) {
        self.provider = provider
        self.id = id
        self.name = name
        self.namespacePath = namespacePath
        let resolvedPath = pathWithNamespace ?? Self.makePath(namespacePath: namespacePath, name: name)
        self.pathWithNamespace = resolvedPath
        self.slug = slug ?? name
        self.webURL = webURL
        self.apiURL = apiURL
        self.providerSpecificID = providerSpecificID
    }

    public static func github(
        id: String,
        owner: String,
        name: String,
        webURL: URL? = nil,
        apiURL: URL? = nil,
        providerSpecificID: String? = nil
    ) -> RepositoryIdentity {
        RepositoryIdentity(
            provider: .github,
            id: id,
            name: name,
            namespacePath: owner,
            webURL: webURL,
            apiURL: apiURL,
            providerSpecificID: providerSpecificID
        )
    }

    private static func makePath(namespacePath: String, name: String) -> String {
        let namespace = namespacePath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard !namespace.isEmpty else { return name }

        return "\(namespace)/\(name)"
    }
}

File: Windows/RepoBar.Core/Models/RepositoryIdentity.cs
namespace RepoBar.Core.Models;

public sealed record RepositoryIdentity
{
    public RepositoryIdentity(
        SourceControlProvider provider,
        string id,
        string name,
        string namespacePath,
        string? pathWithNamespace = null,
        string? slug = null,
        Uri? webUrl = null,
        Uri? apiUrl = null,
        string? providerSpecificId = null)
    {
        Provider = provider;
        Id = id;
        Name = name;
        NamespacePath = namespacePath;
        PathWithNamespace = pathWithNamespace ?? MakePath(namespacePath, name);
        Slug = slug ?? name;
        WebUrl = webUrl;
        ApiUrl = apiUrl;
        ProviderSpecificId = providerSpecificId;
    }

    public SourceControlProvider Provider { get; }

    public string Id { get; }

    public string Name { get; }

    public string NamespacePath { get; }

    public string PathWithNamespace { get; }

    public string Slug { get; }

    public Uri? WebUrl { get; }

    public Uri? ApiUrl { get; }

    public string? ProviderSpecificId { get; }

    public static RepositoryIdentity GitHub(
        string id,
        string owner,
        string name,
        Uri? webUrl = null,
        Uri? apiUrl = null,
        string? providerSpecificId = null) =>
        new(
            SourceControlProvider.GitHub,
            id,
            name,
            owner,
            webUrl: webUrl,
            apiUrl: apiUrl,
            providerSpecificId: providerSpecificId);

    private static string MakePath(string namespacePath, string name)
    {
        string trimmed = namespacePath.Trim('/');
        return string.IsNullOrEmpty(trimmed) ? name : $"{trimmed}/{name}";
    }
}
